src/vsr/superblock_quorums.zig

```diff
@@ -0,0 +1,625 @@
+    try t(2, &.{ m(2), m(2), o(3), o(3) }, 3);
+    try t(2, &.{ m(2), o(3), o(3), o(3) }, 3);
+    // Grandparent has wrong cluster|replica.
+    try t(2, &.{ m(2), m(2), m(2), o(4) }, 2);
+    try t(2, &.{ m(2), m(2), o(4), o(4) }, 4);
+    try t(2, &.{ m(2), o(4), o(4), o(4) }, 4);
+
+    // Parent/child hash chain is broken.
+    const p = CopyTemplate.make_invalid_parent;
+    try t(2, &.{ o(2), o(2), o(2), p(3) }, 2);
+    try t(2, &.{ o(2), o(2), p(3), p(3) }, error.ParentNotConnected);
+    try t(2, &.{ o(2), p(3), p(3), p(3) }, error.ParentNotConnected);
+    try t(2, &.{ p(3), p(3), p(3), p(3) }, 3);
+
+    // Parent view is greater than child view.
+    const v = CopyTemplate.make_invalid_vsr_state;
+    try t(2, &.{ v(2), v(2), o(3), o(3) }, error.VSRStateNotMonotonic);
+
+    // A member of the quorum has an "invalid" copy, but an otherwise valid checksum.
+    const h = CopyTemplate.make_valid_high_copy;
+    try t(2, &.{ o(2), o(2), o(3), h(3) }, 3);
+}
+
+const CopyTemplate = struct {
+    sequence: u64,
+    variant: enum {
+        valid,
+        valid_high_copy,
+        invalid_broken,
+        invalid_misdirect,
+        invalid_parent,
+        invalid_vsr_state,
+    },
+
+    fn make_valid(sequence: u64) CopyTemplate {
+        return .{ .sequence = sequence, .variant = .valid };
+    }
+
+    /// Construct a copy with a corrupt copy index (≥superblock_copies).
+    fn make_valid_high_copy(sequence: u64) CopyTemplate {
+        return .{ .sequence = sequence, .variant = .valid_high_copy };
+    }
+
+    /// Construct a corrupt (invalid checksum) or duplicate copy copy.
+    fn make_invalid_broken(_: void) CopyTemplate {
+        // Use a high sequence so that invalid copies are the last generated by
+        // test_quorums_working(), so that they can become duplicates of (earlier) valid copies.
+        return .{ .sequence = 6, .variant = .invalid_broken };
+    }
+
+    /// Construct a copy with either an incorrect "cluster" or "replica".
+    fn make_invalid_misdirect(sequence: u64) CopyTemplate {
+        return .{ .sequence = sequence, .variant = .invalid_misdirect };
+    }
+
+    /// Construct a copy with an invalid "parent" checksum.
+    fn make_invalid_parent(sequence: u64) CopyTemplate {
+        return .{ .sequence = sequence, .variant = .invalid_parent };
+    }
+
+    /// Construct a copy with a newer `VSRState` than its parent.
+    fn make_invalid_vsr_state(sequence: u64) CopyTemplate {
+        return .{ .sequence = sequence, .variant = .invalid_vsr_state };
+    }
+
+    fn less_than(_: void, a: CopyTemplate, b: CopyTemplate) bool {
+        return a.sequence < b.sequence;
+    }
+};
+
+fn test_quorums_working(
+    threshold_count: u8,
+    copies: *[4]CopyTemplate,
+    result: QuorumsType(.{ .superblock_copies = 4 }).Error!u64,
+) !void {
+    const Quorums = QuorumsType(.{ .superblock_copies = 4 });
+    var prng = std.rand.DefaultPrng.init(@intCast(u64, std.time.milliTimestamp()));
```

Comment:
Do you want to split this like segmented_array_fuzz? Fixed seed for unit tests and a `main` that tries other seeds.

Reply:
Similarly below.